llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp

```diff
@@ -11358,6 +11361,37 @@ bool ARMAsmParser::parseDirectiveARM(SMLoc L) {
   return false;
 }
 
+MCSymbolRefExpr::VariantKind
+ARMAsmParser::getVariantKindForName(StringRef Name) const {
```

Comment:
I can see a design trade-off here. By implementing only the expressions that are supported on Arm we find some problems earlier, however we may miss out on generic expressions added to the `MCSymbolRefExpr::getVariantKindForName` but not for Arm. I can see these getting out of sync.

Would it be worth calling the MCSymbolRefExpr::getVariantKindForName if the Name isn't found? Ideally we could refactor so that MCSymbolRefExpr::getVariantKindForName contains only generic and no target specific variant kinds.

Reply:
> Ideally we could refactor so that MCSymbolRefExpr::getVariantKindForName contains only generic and no target specific variant kinds.

I agree with this. I believe that binutils doesn't have a generic operand modifier, so ideally `MCSymbolRefExpr::getVariantKindForName` should be a pure virtual function. Unfortunately, many architectures add arch-specific modifiers in this list. We should eventually move them to respective lib/Target/XXX files.

In this patch I am only focusing on the ARM specific operand modifiers.